### DISCUSSION TITLE
Get the size of a repo in graphql

### DISCUSSION THREAD

**Discussion by ArivAfonso** [community/community#24648] [Apps, API and Webhooks]
<p>This is essentially  <code>GET /repos/:user/:repo/size</code> but in Graphql. My first thought was to get the size of languages and just add them up together like this</p>
<pre><code class="lang-auto">query { 
  repository(name:"spacex-datasets", owner:"ariv797"){
    languages(first:100){
      edges{
        size
      }
    }
  }
}
</code></pre>
<p>But this does not count things like JSON. So how do you get the size of a repository in Graphql?</p>

**Comment by jfullstackdev** [comment]
<p>how about the API ?</p>
<p><a href="https://docs.github.com/en/rest/reference/repos#get-a-repository" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/rest/reference/repos#get-a-repository</a></p>

**Comment by byrneh** [comment, marked as answer]
<p>Hi <a class="mention" href="/ariv797">@ariv797</a> you solution for question how to get  size of repo GraphQL is below</p>
<pre><code class="lang-auto">query { 
  repository(name:"spacex-datasets", owner:"ariv797"){
    diskUsage
  }
}
</code></pre>

**Reply by psmoot** [reply]
What exactly does that measure? I'm trying to use it and compared to a cloned repo. The value reported by GitHub doesn't match anything I can get with `du` or `ls`.

The docs say it's size on disk in kilobytes (don't know if that's kilo or kibi but no matter). GitHub is reporting 37 and the repo on disk should be closer to 1.5MB (for the whole thing plus worktree) or 900 KB (for just the `.git` directory) or 500k for just the source files.

**Reply by TheNewThinkTank** [reply]
Any way to extend it to fetch say, top 10 largest repo (for logged in user's public repos)?

**Reply by LangLangBart** [reply]
@psmoot
- the official `GraphQL` documentation only says:
> _The number of kilobytes this repository occupies on disk._
> Source: [GitHub Docs - GraphQL API - Objects ](https://docs.github.com/en/graphql/reference/objects)

- in a blog post the size is described as:
> _When people talk about the size of a Git repository, they often talk about the total size needed by Git to store the project’s history in its internal, highly-compressed format—basically, the amount of disk space used by the .git directory. This number is easy to measure. It’s also useful, because it indicates how long it takes to clone the repository and how much disk space it will use._
> Source: [GitHub Blog - Measuring the many sizes of a Git repository](https://github.blog/2018-03-05-measuring-the-many-sizes-of-a-git-repository/) (5/Mar/18)

---

@TheNewThinkTank 

- the `repositories` object does not support sorting them by size through `GraphQL`, the choices are `CREATED_AT`, `UPDATED_AT`, `PUSHED_AT`, `STARGAZERS` and by `NAME`
  - you would need to process the JSON response yourself
  - but it would at least show you the total size of all your repos 

```graphql
{
  repositoryOwner(login: "TheNewThinkTank") {
    repositories(first: 10, orderBy: {field: UPDATED_AT, direction: DESC}, privacy: PUBLIC, isFork: false) {
      totalCount
      totalDiskUsage
      nodes {
        name
        diskUsage
      }
    }
  }
}
```

```json
{
  "data": {
    "repositoryOwner": {
      "repositories": {
        "totalCount": 28,
        "totalDiskUsage": 480483,
        "nodes": [
          {
            "name": "polylang",
            "diskUsage": 43
          },
          {
            "name": "project-metrics",
            "diskUsage": 53
          },
          {
            "name": "security-scanner",
            "diskUsage": 56
          },
          {
            "name": "interactive-musicology",
            "diskUsage": 1
          },
          {
            "name": "chess-explorer",
            "diskUsage": 1
          },
          {
            "name": "TheNewThinkTank",
            "diskUsage": 7402
          },
          {
            "name": "Illustris-Zoom-Simulation",
            "diskUsage": 2395
          },
          {
            "name": "Hypothesis-Testing",
            "diskUsage": 118
          },
          {
            "name": "Web-Application-jQuery-And-Bootstrap",
            "diskUsage": 14761
          },
          {
            "name": "Modelling-And-Computations-MATLAB",
            "diskUsage": 11984
          }
        ]
      }
    }
  }
}
```

---

- alternatively you could utilize the `search` object and make use of valid `qualifiers`, for example, lets search the repos of a specific user with a size larger than 1000kb, that are public and aren't forks.
  - see link with valid `qualifiers`: [GitHub Docs - Searching on GitHub](https://docs.github.com/en/search-github/searching-on-github)

```graphql
{
  search(query: "user:TheNewThinkTank size:>1000 is:public fork:false", type: REPOSITORY, first: 10) {
    repositoryCount
    nodes {
      ... on Repository {
        name
        diskUsage
      }
    }
  }
}
```

```json
{
  "data": {
    "search": {
      "repositoryCount": 10,
      "nodes": [
        {
          "name": "Fitness-Tracker",
          "diskUsage": 10226
        },
        {
          "name": "N-Body-Simulations",
          "diskUsage": 32266
        },
        {
          "name": "Illustris-Zoom-Simulation",
          "diskUsage": 2395
        },
        {
          "name": "dark-matter-attractor",
          "diskUsage": 55377
        },
        {
          "name": "Modelling-And-Computations-MATLAB",
          "diskUsage": 11984
        },
        {
          "name": "Web-Application-jQuery-And-Bootstrap",
          "diskUsage": 14761
        },
        {
          "name": "AACT-Analysis",
          "diskUsage": 326467
        },
        {
          "name": "image-models",
          "diskUsage": 16560
        },
        {
          "name": "twitter-novo",
          "diskUsage": 1049
        },
        {
          "name": "TheNewThinkTank",
          "diskUsage": 7402
        }
      ]
    }
  }
}
```

**Reply by TheNewThinkTank** [reply]
Amazing, exactly what I needed! Thank you @LangLangBart
